tools/brokerd/service/win32/broker_service.cpp

```diff
@@ -36,35 +44,53 @@ sample_service::~sample_service () {
 /// begun: it must not block.
 
-void sample_service::start_handler (DWORD /*argc*/, TCHAR * /*argv*/[]) {
-    // Log a service start message to the Application log.
-    this->write_event_log_entry ("CppWindowsService in OnStart", event_type::information);
-    thread_ = std::thread ([this] () { this->worker (); });
-}
+void broker_service::start_handler (DWORD argc, TCHAR * argv[]) {
+    this->write_event_log_entry ("broker service starting", event_type::information);
+
+    command_line::option::reset_container ();
 
+    switches opt;
+    std::tie (opt, broker::exit_code) = get_switches (argc, argv);
+
+    if (broker::exit_code != EXIT_SUCCESS) {
+        this->write_event_log_entry ("error: broker service failed to parse commandline options",
+                                     event_type::error);
+        this->set_service_status (SERVICE_STOPPED, EXIT_FAILURE);
+        return;
+    }
+
+    this->worker_thread_ = std::thread ([this, opt] { this->worker (opt); });
+
+    this->write_event_log_entry ("broker service started successfully", event_type::information);
+}
 
 // worker
 // ~~~~~~
-/// The method performs the main function of the service. It runs on a thread pool worker thread.
-void sample_service::worker () {
-    // Periodically check if the service is stopping.
-    while (!is_stopping_) {
-        // Perform main service function here...
-        using namespace std::chrono_literals;
-
-        wprintf (TEXT ("Stopping %s."), this->name ().c_str ());
-        std::this_thread::sleep_for (2s); // Simulate some lengthy operations.
+void broker_service::worker (switches opt) {
+    try {
+        broker::run_broker (opt);
+        if (broker::exit_code != EXIT_SUCCESS) {
+            this->write_event_log_entry ("broker service exited unsuccessfully", event_type::error);
+        }
+    } catch (std::exception const & ex) {
+        this->write_event_log_entry (std::string ("error: ", ex.what ()).c_str (),
+                                     event_type::error);
+        this->set_service_status (SERVICE_STOPPED, EXIT_FAILURE);
+    } catch (...) {
+        this->write_event_log_entry ("unknown error!", event_type::error);
+        this->set_service_status (SERVICE_STOPPED, EXIT_FAILURE);
     }
 }
 
 // stop_handler
 // ~~~~~~~~~~~~
 /// \note Periodically call ReportServiceStatus() with SERVICE_STOP_PENDING if the procedure is
 /// going to take long time.
-void sample_service::stop_handler () {
-    this->write_event_log_entry ("CppWindowsService in stop_handler", event_type::information);
-
-    // Indicate that the service is stopping and join the thread.
-    is_stopping_ = true;
-    thread_.join ();
+void broker_service::stop_handler () {
+    this->write_event_log_entry ("broker quiting", event_type::information);
+    broker::notify_quit_thread ();
+    while (!this->worker_thread_.joinable()) {
+        this->set_service_status (SERVICE_STOP_PENDING);
+    }
```

Comment:
I’m not too keen on this spin-loop. service_base::stop() sets the service status to SERVICE_STOP_PENDING before calling stop_handler(); is this enough? If not, then this loop would be better waiting for a the worker thread to notify a condition variable that it is about to exit.

Reply:
I think it's alright to not have the loop and the call to `set_service_status()` and instead just directly call `join()` on the thread. It still causes a clean shutdown and is fine from what I could read.